### ISSUE TITLE
Android Sentry not sending exceptions that created under Thread.setDefaultUncaughtExceptionHandler

### ISSUE THREAD

**Issue by ndasim** [getsentry/sentry-java#953]
Let me start with our use case: Our application needs to restart itself if any unhandled exception occurs, this feature crucial for most of our customers. While handling unhandled exception our application attempting to send crash data with sentry and restarts itself with the help of AlarmManager

Before upgrading our sentry library from `com.getsentry.raven:raven-android:8.0.2` to `io.sentry:sentry-android:2.2.1`, the following code snipped was working for us to the sending those crashes:
```
EventBuilder eventBuilder = new EventBuilder()
        .withMessage(tag)
        .withLevel(Event.Level.ERROR)
        .withSentryInterface(new ExceptionInterface(e));

Raven.getStoredInstance().runBuilderHelpers(eventBuilder);
Raven.getStoredInstance().capture(eventBuilder.build());
```

But now, since we changed the sentry library, the following snipped is not sending exception:
```
Sentry.withScope(scope -> {
    scope.addEventProcessor((event, hint) -> {
        Message message = new Message();
        message.setMessage("CRASH");
        event.setLevel(SentryLevel.FATAL);
        event.setMessage(message);
        return event;
    });

    Sentry.captureException(e);
});
```

This code piece working in Thread.setDefaultUncaughtExceptionHandler(Thread.UncaughtExceptionHandler handler)'s handler, after executing snipped I'm using AlarmManager to re-launch the destroyed application.

Other build variables that you might need to know:
```
compileSdkVersion 22
buildToolsVersion '28.0.3'
minSdkVersion 21
targetSdkVersion 22
multiDexEnabled true
minifyEnabled true
```
Also sentry android gradle plugin version is 1.7.28.

Maybe I lost in documentations and I lost some key points, but the new snipped is working only on the debug build. It's only not working on the release build. To be honest, during my tests, it sent the crash event only once :D after 10-20 crash attempt. I thought that sentry couldn't be able to send an event due to fast restart and I tried Sentry.flush() and hopelessly tried Thread.sleep() to slow down the restart procedure. Also, I tried changing AlarmManager's relaunch duration 1 second to 5-6 seconds. None of them worked for me.   

I'm open for your kind response and suggestions. Thank you

**Comment by marandaneto** [comment]
@ndasim thanks for your question.

you don't need to `Thread .setDefaultUncaughtExceptionHandler` yourself, the SDK does it automatically already.
Also, you don't need to `Sentry.withScope` nor `scope.addEventProcessor`, why are you trying to do this? just to understand your use case.

If you have an exception, just do `Sentry.captureException(e)`
If you want to capture a message, just do `Sentry.captureMessage(message)` or `Sentry.captureEvent(...)` if you want a custom event.

If you want to capture unhandled exception (hardcrash), you don't need to do anything, just let your App to crash and the SDK will do the rest on App. restart.

**Comment by ndasim** [comment]
@marandaneto Thanks for the response, I need  `scope.addEventProcessor`  with `Sentry.withScope` to attach some internal session data for analyzing impact of those issues, actually I removed that part from snipped while sharing with you. 

Does `Sentry.withScope` affecting my scenario? We strongly need to attach extra information for those exceptions. I need to attach some tags and some internal usage breadcrumbs which contains custom events.

If are there any other possible way for this case I would be appreciated.

**Comment by marandaneto** [comment]
@ndasim yes.
Look at the `Sentry` static class.
`Sentry.setTag(...)`, `Sentry.addBreadcrumb(...)`, etc... on Mobile there's a single shared Hub/Scope, if you add a tag and/or breadcrumb, every `Sentry.captureException(e)` will contain all that info on the event automatically.

**Comment by ndasim** [comment]
Okey, I tested your suggestion but did not work for me. Then I simply put 
```
try {
    Thread.sleep(1000);
} catch (InterruptedException ex) {
    ex.printStackTrace();
}
```
Thread sleep code piece under my `addEventProcessor` definition. 100ms delay is working great, in case of any thread overloading case on the Sentry itself, I increased it to 1 second. Basically, I'm postponing the application shutdown for the jobs of other threads.

And it worked! Actually, I have tried that method before but during my tests but I realized that I was inspecting the wrong environment :D because of our flavor based configurations. 

I need to attach that, I was using `SentryAndroid.init` with `options.setBeforeSend` that might help you understand the issue and fix it on the docs or next release.

I'm closing with this ticket with appreciation.
